File: 31.下一个排列.cs
/*
 * @lc app=leetcode.cn id=31 lang=csharp
 *
 * [31] 下一个排列
 */

// @lc code=start
public class Solution
{
    public void NextPermutation(int[] nums)
    {
        int len = nums.Length;
        int left = len - 2;
        int right = len - 1;
        for (; left >= 0 && nums[left] >= nums[left + 1]; left--) { }
        if (left >= 0)
        {
            for (; right >= 0 && nums[left] >= nums[right]; right--) { }
            nums[left] ^= nums[right];
            nums[right] ^= nums[left];
            nums[left] ^= nums[right];
        }
        ReverseNums(nums, left + 1);
        return;
    }

    public void ReverseNums(int[] nums, int left)
    {
        int right = nums.Length - 1;
        while (left < right)
        {
            nums[left] ^= nums[right];
            nums[right] ^= nums[left];
            nums[left] ^= nums[right];
            left++;
            right--;
        }
        return;
    }
}
// @lc code=end

File: 31.下一个排列.rs
/*
 * @lc app=leetcode.cn id=31 lang=rust
 *
 * [31] 下一个排列
 */

// @lc code=start
impl Solution {
    pub fn next_permutation(nums: &mut Vec<i32>) {
        let len = nums.len() as i32;
        let mut left = len - 2;
        let mut right = len - 1;
        while left >= 0 && nums[left as usize] >= nums[left as usize + 1] {
            left -= 1;
        }
        if left >= 0 {
            while right >= 0 && nums[left as usize] >= nums[right as usize] {
                right -= 1;
            }
            (nums[left as usize],nums[right as usize]) = (nums[right as usize],nums[left as usize]);
        }
        Solution::reverted_nums(nums, left as usize + 1)
    }
    
    pub fn reverted_nums(nums: &mut Vec<i32>, mut left: usize) {
        let mut right = nums.len() - 1;
        while left < right {
            (nums[left as usize],nums[right as usize]) = (nums[right as usize],nums[left as usize]);
            left += 1;
            right -= 1;
        }
    }
}
// @lc code=end

